File: src/qt5_gui/mainwindow.cpp
#define NO_USING_QTYPES
#include <common.h>

#include "mainwindow.hpp"

#include <QFontDatabase>
#include <QPlainTextEdit>
#include <QtWidgets>
#include <qsys/command/CmdMgr.hpp>
// #include <qsys/command/Command.hpp>
#include <pybr/PythonBridge.hpp>
#include <pybr/pybr.hpp>

#include "QtMolStructPanel.hpp"
#include "QtMolWidget2.hpp"
#include "QtScenePanel.hpp"
#include "moc_mainwindow.cpp"
#include "qlib/LMsgLog.hpp"
#include "qsys/SceneManager.hpp"

MainWindow::MainWindow()
{
    createWidgets();

    createActions();
    createStatusBar();
    // createDockWindows();
    updateMenus();

    readSettings();

    setUnifiedTitleAndToolBarOnMac(true);

    auto pLogMgr = qlib::LMsgLog::getInstance();
    auto msg = pLogMgr->getAccumMsg();
    pLogMgr->removeAccumMsg();
    m_pLogWnd->appendPlainText(msg.c_str());
    m_nLogListenerID = pLogMgr->addListener(this);
}

MainWindow::~MainWindow()
{
    auto pLogMgr = qlib::LMsgLog::getInstance();
    pLogMgr->removeListener(m_nLogListenerID);
}

void MainWindow::onLoaded()
{
    if (m_pTabWnd->currentSubWindow() == nullptr) newScene();
}

void MainWindow::onActivateMolTabChanged()
{
    auto pScMgr = qsys::SceneManager::getInstance();
    auto activewnd = m_pTabWnd->activeSubWindow();
    if (activewnd == nullptr) {
        LOG_DPRINTLN("XXXX MainWindow::onActivateMolTabChanged(): deactivated");
        // pScMgr->getActiveSceneID(qlib::invalid_uid);
        return;
    }
    auto molw = activewnd->findChild<QtMolWidget2 *>();
    auto scid = molw->getSceneID();
    auto vwid = molw->getViewID();
    auto title = activewnd->windowTitle();
    printf("XXXX MainWindow::onActivateMolTabChanged(%s:s%d:v%d)\n",
           title.toUtf8().constData(), scid, vwid);
    fflush(stdout);

    auto pActSc = pScMgr->getScene(scid);
    pActSc->setActiveViewID(vwid);
    // pScMgr->getActiveSceneID(scid);
}

void MainWindow::createWidgets()
{
    m_pTabWnd = new QMdiArea;
    m_pTabWnd->setHorizontalScrollBarPolicy(Qt::ScrollBarAsNeeded);
    m_pTabWnd->setVerticalScrollBarPolicy(Qt::ScrollBarAsNeeded);
    m_pTabWnd->setViewMode(QMdiArea::TabbedView);
    m_pTabWnd->setTabsClosable(true);
    m_pTabWnd->setTabsMovable(true);
    QTabBar *pBar = m_pTabWnd->findChild<QTabBar *>();
    pBar->setExpanding(false);

    // connect(m_pTabWnd, &QMdiArea::subWindowActivated, this,
    //         &MainWindow::onActivateMolTabChanged);
    connect(m_pTabWnd, SIGNAL(subWindowActivated(QMdiSubWindow *)), this,
            SLOT(onActivateMolTabChanged()));

    //////////

    // QDockWidget *dock = new QDockWidget(tr("Scene"), this);
    auto &&pdock1 = new QtScenePanel();
    addDockWidget(Qt::LeftDockWidgetArea, pdock1);

    auto &&pdock2 = new QtMolStructPanel();
    addDockWidget(Qt::LeftDockWidgetArea, pdock2);

    //////////

    m_pLogWnd = new QPlainTextEdit();
    m_pLogWnd->setReadOnly(true);
    m_pLogWnd->setMinimumSize(1, 1);
    auto &&fixedFont = QFontDatabase::systemFont(QFontDatabase::FixedFont);
    m_pLogWnd->setFont(fixedFont);

    m_pSplitter = new QSplitter(Qt::Vertical);
    // m_pSplitter->addWidget(m_pMolWidget);
    m_pSplitter->addWidget(m_pTabWnd);
    m_pSplitter->addWidget(m_pLogWnd);

    m_pSplitter->setStretchFactor(0, 20);
    m_pSplitter->setStretchFactor(1, 1);

    // setCentralWidget(m_pMolWidget);
    setCentralWidget(m_pSplitter);
}

void MainWindow::logAppended(qlib::LLogEvent &evt)
{
    auto &&msg = evt.getMessage();
    m_pLogWnd->appendPlainText(msg.c_str());
    m_pLogWnd->verticalScrollBar()->setValue(m_pLogWnd->verticalScrollBar()->maximum());
}

//////////

void MainWindow::closeEvent(QCloseEvent *event)
{
    m_pTabWnd->closeAllSubWindows();
    if (m_pTabWnd->currentSubWindow()) {
        event->ignore();
    } else {
        writeSettings();
        event->accept();
    }
}

void MainWindow::newScene()
{
    auto pMgr = qsys::CmdMgr::getInstance();
    pMgr->runGUICmd("qt_new_scene", this);
}

void MainWindow::openScene()
{
    auto pMgr = qsys::CmdMgr::getInstance();
    pMgr->runGUICmd("qt_load_scene", this);
}

void MainWindow::openObject()
{
    auto pMgr = qsys::CmdMgr::getInstance();
    pMgr->runGUICmd("qt_load_object", this);
}

void MainWindow::execPyScr()
{
    QFileDialog dlg(this, "Open python script", "", "");
    if (dlg.exec() != QDialog::Accepted) {
        return;
    }
    auto files = dlg.selectedFiles();
    pybr::PythonBridge *pSvc = pybr::PythonBridge::getInstance();
    for (const auto &f : files) {
        auto filePath = LString(f.toUtf8().constData());
        LOG_DPRINTLN("Selected file: %s", filePath.c_str());
        try {
            pSvc->runFile(filePath);
        } catch (const qlib::LException &e) {
            LOG_DPRINTLN("Caught exception <%s>", typeid(e).name());
            LOG_DPRINTLN("Reason: %s", e.getMsg().c_str());
        } catch (std::exception &e) {
            LOG_DPRINTLN("Caught exception <%s>", typeid(e).name());
            LOG_DPRINTLN("Reason: %s", e.what());
        } catch (...) {
            LOG_DPRINTLN("Caught unknown exception");
        }
    }
}

bool MainWindow::openFile(const QString &fileName)
{
    const bool succeeded = loadFile(fileName);
    if (succeeded) statusBar()->showMessage(tr("File loaded"), 2000);
    return succeeded;
}

bool MainWindow::loadFile(const QString &fileName)
{
    auto &&pchild = createMolWidget();
    pchild->createSceneAndView();

    pchild->loadFile(fileName);
    pchild->showMaximized();
    pchild->update();

    // TO DO: close on failure
    return true;
}

static inline QString recentFilesKey()
{
    return QStringLiteral("recentFileList");
}
static inline QString fileKey()
{
    return QStringLiteral("file");
}

static QStringList readRecentFiles(QSettings &settings)
{
    QStringList result;
    const int count = settings.beginReadArray(recentFilesKey());
    for (int i = 0; i < count; ++i) {
        settings.setArrayIndex(i);
        result.append(settings.value(fileKey()).toString());
    }
    settings.endArray();
    return result;
}

static void writeRecentFiles(const QStringList &files, QSettings &settings)
{
    const int count = files.size();
    settings.beginWriteArray(recentFilesKey());
    for (int i = 0; i < count; ++i) {
        settings.setArrayIndex(i);
        settings.setValue(fileKey(), files.at(i));
    }
    settings.endArray();
}

bool MainWindow::hasRecentFiles()
{
    QSettings settings(QCoreApplication::organizationName(),
                       QCoreApplication::applicationName());
    const int count = settings.beginReadArray(recentFilesKey());
    settings.endArray();
    return count > 0;
}

void MainWindow::prependToRecentFiles(const QString &fileName)
{
    QSettings settings(QCoreApplication::organizationName(),
                       QCoreApplication::applicationName());

    const QStringList oldRecentFiles = readRecentFiles(settings);
    QStringList recentFiles = oldRecentFiles;
    recentFiles.removeAll(fileName);
    recentFiles.prepend(fileName);
    if (oldRecentFiles != recentFiles) writeRecentFiles(recentFiles, settings);

    setRecentFilesVisible(!recentFiles.isEmpty());
}

void MainWindow::setRecentFilesVisible(bool visible)
{
    recentFileSubMenuAct->setVisible(visible);
    recentFileSeparator->setVisible(visible);
}

void MainWindow::updateRecentFileActions()
{
    QSettings settings(QCoreApplication::organizationName(),
                       QCoreApplication::applicationName());

    const QStringList recentFiles = readRecentFiles(settings);
    const int count = qMin(int(MaxRecentFiles), recentFiles.size());
    int i = 0;
    for (; i < count; ++i) {
        const QString fileName = QFileInfo(recentFiles.at(i)).fileName();
        recentFileActs[i]->setText(tr("&%1 %2").arg(i + 1).arg(fileName));
        recentFileActs[i]->setData(recentFiles.at(i));
        recentFileActs[i]->setVisible(true);
    }
    for (; i < MaxRecentFiles; ++i) recentFileActs[i]->setVisible(false);
}

void MainWindow::openRecentFile()
{
    if (const QAction *action = qobject_cast<const QAction *>(sender()))
        openFile(action->data().toString());
}

void MainWindow::save() {}

void MainWindow::saveAs() {}

#ifndef QT_NO_CLIPBOARD
void MainWindow::cut() {}

void MainWindow::copy() {}

void MainWindow::paste() {}
#endif

void MainWindow::about()
{
    QMessageBox::about(this, tr("About CueMol"), tr("CueMol GUI"));
}

void MainWindow::updateMenus() {}

void MainWindow::updateWindowMenu()
{
    windowMenu->clear();
    windowMenu->addAction(closeAct);
    windowMenu->addAction(closeAllAct);
    windowMenu->addSeparator();
    windowMenu->addAction(tileAct);
    windowMenu->addAction(cascadeAct);
    windowMenu->addSeparator();
    windowMenu->addAction(nextAct);
    windowMenu->addAction(previousAct);
    windowMenu->addAction(windowMenuSeparatorAct);
}

void MainWindow::createActions()
{
    QMenu *fileMenu = menuBar()->addMenu(tr("&File"));
    fileMenu->setObjectName("fileMenu"); 
    auto *p = menuBar()->findChild<QMenu*>("fileMenu");
    LOG_DPRINTLN("******* %p %p", fileMenu, p);
    // QToolBar *fileToolBar = addToolBar(tr("File"));
    // fileToolBar->setObjectName("fileToolBar");

    // File-New action

    // const QIcon newIcon = QIcon::fromTheme("document-new",
    // QIcon(":/images/new.png")); newAct = new QAction(newIcon, tr("&New"), this);
    newAct = new QAction(tr("&New"), this);
    newAct->setShortcuts(QKeySequence::New);
    newAct->setStatusTip(tr("Create a new file"));
    connect(newAct, &QAction::triggered, this, &MainWindow::newScene);
    fileMenu->addAction(newAct);
    // fileToolBar->addAction(newAct);

    // File-Open scene action

    // const QIcon openIcon = QIcon::fromTheme("document-open",
    // QIcon(":/images/open.png")); QAction *openSceneAct = new QAction(openIcon,
    // tr("&Open..."), this);
    QAction *openSceneAct = new QAction(tr("&Open scene..."), this);
    openSceneAct->setShortcuts(QKeySequence::Open);
    openSceneAct->setStatusTip(tr("Open an existing scene file"));
    connect(openSceneAct, &QAction::triggered, this, &MainWindow::openScene);
    fileMenu->addAction(openSceneAct);
    // fileToolBar->addAction(openSceneAct);

    // File-Open object action
    QAction *openObjAct = new QAction(tr("&Open file..."), this);
    openObjAct->setShortcuts(QKeySequence::Open);
    openObjAct->setStatusTip(tr("Open an existing object file"));
    connect(openObjAct, &QAction::triggered, this, &MainWindow::openObject);
    fileMenu->addAction(openObjAct);

    // File-Open python script action
    QAction *openPyScrAct = new QAction(tr("&Exec pyscr..."), this);
    // openPyScrAct->setShortcuts(QKeySequence::Open);
    openPyScrAct->setStatusTip(tr("Execute python script"));
    connect(openPyScrAct, &QAction::triggered, this, &MainWindow::execPyScr);
    fileMenu->addAction(openPyScrAct);

    const QIcon exitIcon = QIcon::fromTheme("application-exit");
    QAction *exitAct = fileMenu->addAction(exitIcon, tr("E&xit"), qApp,
                                           &QApplication::closeAllWindows);
    exitAct->setShortcuts(QKeySequence::Quit);
    exitAct->setStatusTip(tr("Exit the application"));
    fileMenu->addAction(exitAct);
}

void MainWindow::createStatusBar()
{
    statusBar()->showMessage(tr("Ready"));
}

void MainWindow::readSettings()
{
    QSettings settings(QCoreApplication::organizationName(),
                       QCoreApplication::applicationName());
    const QByteArray geometry = settings.value("geometry", QByteArray()).toByteArray();

    if (geometry.isEmpty()) {
        const QRect availableGeometry = screen()->availableGeometry();
        resize(availableGeometry.width() / 3, availableGeometry.height() / 2);
        move((availableGeometry.width() - width()) / 2,
             (availableGeometry.height() - height()) / 2);
    } else {
        restoreGeometry(geometry);
    }

    const QByteArray state = settings.value("windowState", QByteArray()).toByteArray();
    if (!state.isEmpty()) {
        restoreState(state);
    }
}

void MainWindow::writeSettings()
{
    QSettings settings(QCoreApplication::organizationName(),
                       QCoreApplication::applicationName());
    settings.setValue("geometry", saveGeometry());
    settings.setValue("windowState", saveState());
}

void MainWindow::switchLayoutDirection()
{
    if (layoutDirection() == Qt::LeftToRight)
        QGuiApplication::setLayoutDirection(Qt::RightToLeft);
    else
        QGuiApplication::setLayoutDirection(Qt::LeftToRight);
}

void MainWindow::showEvent(QShowEvent *event)
{
    QWidget::showEvent(event);
    QTimer::singleShot(0, this, &MainWindow::onLoaded);
}

QtMolWidget2 *MainWindow::createMolWidget()
{
    auto *pchild = new QtMolWidget2();
    pchild->setObjectName("MolWidget");
    m_pTabWnd->addSubWindow(pchild);
    return pchild;
}

QtMolWidget2 *MainWindow::activeMolWidget()
{
    auto *pActSubWnd = m_pTabWnd->activeSubWindow();
    if (pActSubWnd == nullptr) return nullptr;
    return pActSubWnd->findChild<QtMolWidget2 *>();
}

File: src/python/cuemol_gui/main_window.py
import json

from cuemol_gui.event_manager import EventManager
from cuemol_gui.gui_command_manager import GUICommandManager
from PySide2 import QtCore, QtWidgets
from PySide2.QtCore import QSettings, Qt
# from PySide2.QtWidgets import QFileDialog
from PySide2.QtGui import QFont, QIcon
# from PySide2.QtOpenGL import QGLFormat
from PySide2.QtWidgets import QAction, QApplication, QMainWindow, QMdiArea, QTabBar
from qt5gui import QtMolWidget

import cuemol

# import main


# from main import event, CmdPrompt


class MainWindow(QMainWindow):
    def __init__(self, parent=None):
        # qt5gui.qt5gui_init()
        super().__init__(parent)
        self.init_ui()
        self.show()
        self.on_new_scene()

    def closeEvent(self, event):
        print("MainWindow.closeEvent called!!")
        self.save_settings()

    def on_exit_app(self):
        print("MainWindow.on_exit_app called!!")
        self.save_settings()
        QApplication.instance().quit()

    def on_log_event(self, aSlotID, aCatStr, aTgtTypeID, aEvtTypeID, aSrcID, aInfoStr):
        info = json.loads(str(aInfoStr))
        self.append_log(info["content"])

    def append_log(self, msg):
        self._logwnd.appendPlainText(msg)
        self._logwnd.verticalScrollBar().setValue(
            self._logwnd.verticalScrollBar().maximum()
        )

    def init_ui(self):

        self.create_widgets()
        self.create_menu()

        self.statusBar()

        self.resize(800, 600)
        self.load_settings()
        self.setWindowTitle("CueMol")

        # Setup log event handler
        log_mgr = cuemol.getService("MsgLog")
        accum_msg = log_mgr.getAccumMsg()
        log_mgr.removeAccumMsg()
        self.append_log(accum_msg)
        evm = EventManager.get_instance()
        evm.add_listener(
            "log",
            evm.impl.SEM_ANY,  # source type
            evm.impl.SEM_LOG,  # event type
            evm.impl.SEM_ANY,  # source uid
            self.on_log_event,
        )

    def create_menu(self):
        menubar = self.menuBar()

        # File menu

        scene_open_act = QAction("&Open scene", self)
        scene_open_act.setShortcut("Ctrl+Shift+O")
        scene_open_act.setStatusTip("Open scene")
        scene_open_act.triggered.connect(self.on_open_scene)

        obj_open_act = QAction("&Open file", self)
        obj_open_act.setShortcut("Ctrl+O")
        obj_open_act.setStatusTip("Open file")
        obj_open_act.triggered.connect(self.on_open_object)

        new_scene_act = QAction("&New scene", self)
        new_scene_act.setShortcut("Ctrl+N")
        new_scene_act.setStatusTip("New scene")
        new_scene_act.triggered.connect(self.on_new_scene)

        exit_act = QAction(QIcon("exit.png"), "&Exit", self)
        exit_act.setShortcut("Ctrl+Q")
        exit_act.setStatusTip("Exit application")
        exit_act.triggered.connect(self.on_exit_app)

        fileMenu = menubar.addMenu("&File")
        fileMenu.addAction(scene_open_act)
        fileMenu.addAction(new_scene_act)
        fileMenu.addAction(obj_open_act)
        fileMenu.addAction(exit_act)

        # Edit menu

        undo_act = QAction("&Undo", self)
        undo_act.setShortcut("Ctrl+Z")
        undo_act.setStatusTip("Undo")
        undo_act.triggered.connect(self.on_undo)
        self._undo_act = undo_act

        redo_act = QAction("&Redo", self)
        redo_act.setShortcut("Ctrl+Shift+Z")
        redo_act.setStatusTip("Redo")
        redo_act.triggered.connect(self.on_redo)
        self._redo_act = redo_act

        editMenu = menubar.addMenu("&Edit")
        editMenu.addAction(undo_act)
        editMenu.addAction(redo_act)

        editMenu.aboutToShow.connect(self.on_edit_menu_showing)

    def get_undo_size(self):
        scene, _ = self.active_scene_view()
        print(f"undo scene: {scene}")
        if scene is None:
            return 0
        nundo = scene.getUndoSize()
        print(f"undo size: {nundo}")
        return nundo

    def get_top_undo_description(self):
        scene, _ = self.active_scene_view()
        print(f"undo scene: {scene}")
        if scene is None:
            return None
        undo_desc = scene.getUndoDesc(0)
        print(f"undo desc: {undo_desc}")
        return undo_desc

    def get_redo_size(self):
        scene, _ = self.active_scene_view()
        if scene is None:
            return 0
        nredo = scene.getRedoSize()
        return nredo

    def get_top_redo_description(self):
        scene, _ = self.active_scene_view()
        print(f"redo scene: {scene}")
        if scene is None:
            return None
        redo_desc = scene.getRedoDesc(0)
        print(f"redo desc: {redo_desc}")
        return redo_desc

    def on_edit_menu_showing(self):
        print("on_edit_menu_showing triggered")
        # Update undo menu
        n_undo = self.get_undo_size()
        if n_undo > 0:
            self._undo_act.setEnabled(True)
            desc = self.get_top_undo_description()
            if desc is not None:
                self._undo_act.setText(f"&Undo: {desc}")
            else:
                self._undo_act.setText("&Undo")
        else:
            self._undo_act.setEnabled(False)
            self._undo_act.setText("&Undo")

        # Update redo menu
        n_redo = self.get_redo_size()
        if n_redo > 0:
            self._redo_act.setEnabled(True)
            desc = self.get_top_redo_description()
            if desc is not None:
                self._redo_act.setText(f"&Redo: {desc}")
            else:
                self._redo_act.setText("&Redo")
        else:
            self._redo_act.setEnabled(False)
            self._redo_act.setText("&Redo")

    def active_mol_widget(self):
        active_wnd = self._mdi_area.activeSubWindow()
        print(f"active_wnd: {active_wnd}")
        if active_wnd is None:
            return None
        mol_widget = active_wnd.findChild(QtMolWidget)
        print(f"active mol_widget: {mol_widget}")
        return mol_widget

    def active_scene_view_ids(self):
        mol_widget = self.active_mol_widget()
        if mol_widget is None:
            return None, None
        scid = mol_widget.getSceneID()
        vwid = mol_widget.getViewID()
        print(f"active : {scid}, {vwid}")
        return scid, vwid

    def active_scene_view(self):
        scid, vwid = self.active_scene_view_ids()
        if scid is None:
            return None, None
        mgr = cuemol.svc("SceneManager")
        return mgr.getScene(scid), mgr.getScene(vwid)

    def on_active_moltab_changed(self):
        print("onActiveMolTabChanged called!!")
        scid, vwid = self.active_scene_view_ids()
        if scid is None:
            print("XXXX MainWindow::onActivateMolTabChanged(): deactivated")
            return

        print(f"MainWindow::onActivateMolTabChanged(sc:{scid} vw:{vwid})")
        sc_mgr = cuemol.svc("SceneManager")
        active_scene = sc_mgr.getScene(scid)
        active_scene.setActiveViewID(vwid)

        evm = EventManager.get_instance()
        evm.add_listener(
            "mouseClicked",
            evm.impl.SEM_INDEV,  # target type
            evm.impl.SEM_ANY,  # event type
            vwid,  # source uid
            self.on_molview_clicked,
        )

    def create_widgets(self):
        # Create tabbed mol view container
        mdi_area = QMdiArea()
        mdi_area.setHorizontalScrollBarPolicy(Qt.ScrollBarAsNeeded)
        mdi_area.setVerticalScrollBarPolicy(Qt.ScrollBarAsNeeded)
        mdi_area.setViewMode(QMdiArea.TabbedView)
        mdi_area.setTabsClosable(True)
        mdi_area.setTabsMovable(True)
        # Change tabbar style
        tabbar = mdi_area.findChild(QTabBar)
        tabbar.setExpanding(False)

        # Listen tab events
        mdi_area.subWindowActivated.connect(self.on_active_moltab_changed)
        self._mdi_area = mdi_area

        # Create log widget
        logwnd = QtWidgets.QPlainTextEdit(self)
        logwnd.setReadOnly(True)
        f = QFont("Monospace")
        f.setStyleHint(QFont.TypeWriter)
        logwnd.setFont(f)
        # print("Logwnd minimumSizeHint=" + str( self._logwnd.minimumSizeHint() ))
        logwnd.setMinimumSize(1, 1)
        self._logwnd = logwnd

        # Create molview/logwnd splitter
        splitter = QtWidgets.QSplitter(QtCore.Qt.Vertical)
        splitter.addWidget(mdi_area)
        splitter.addWidget(logwnd)
        # splitter.addWidget(loggrp)
        # splitter.setStretchFactor(0, 20)
        # splitter.setStretchFactor(1, 1)
        self._splitter = splitter

        # Set central widget
        self.setCentralWidget(splitter)

    def create_mol_widget(self):
        mol_widget = QtMolWidget()
        self._mdi_area.addSubWindow(mol_widget)
        print(f"create_mol_widget mol widget: {mol_widget}")
        print(f"create_mol_widget mdi area: {self._mdi_area}")
        # self.active_mol_widget()
        return mol_widget

    def on_molview_clicked(self, aSlotID, aCatStr, aTgtTypeID, aEvtTypeID, aSrcID, aInfoStr):
        print("on_molview_clicked")

    def on_new_scene(self):
        mgr = GUICommandManager.get_instance()
        mgr.run_command("qt_new_scene", self)

    def on_open_scene(self):
        mgr = GUICommandManager.get_instance()
        mgr.run_command("qt_load_scene", self)

    def on_open_object(self):
        mgr = GUICommandManager.get_instance()
        mgr.run_command("qt_load_object", self)

    def on_undo(self):
        scene, _ = self.active_scene_view()
        print(f"undo scene: {scene}")
        if scene is None:
            return
        print(f"undo size: {scene.getUndoSize()}")
        scene.undo(0)

    def on_redo(self):
        scene, _ = self.active_scene_view()
        print(f"redo scene: {scene}")
        if scene is None:
            return
        print(f"redo size: {scene.getRedoSize()}")
        scene.redo(0)

    def save_settings(self):
        # qset = QSettings("BKR-LAB", "CueMol")
        qset = QSettings()
        qset.beginGroup("mainwindow")

        qset.setValue("geometry", self.saveGeometry())
        qset.setValue("savestate", self.saveState())
        qset.setValue("maximized", self.isMaximized())
        print(f"is_maximized: {self.isMaximized()}")
        if not self.isMaximized():
            qset.setValue("pos", self.pos())
            qset.setValue("size", self.size())

        # splitter state
        qset.setValue("splitterSize", self._splitter.saveState())

        qset.endGroup()

    def load_settings(self):
        # qset = QSettings("BKR-LAB", "CueMol")
        qset = QSettings()
        qset.beginGroup("mainwindow")

        self.restoreGeometry(qset.value("geometry", self.saveGeometry()))
        self.restoreState(qset.value("savestate", self.saveState()))
        self.move(qset.value("pos", self.pos()))
        self.resize(qset.value("size", self.size()))

        str_maximized = qset.value("maximized")
        print(f"str_maximized: {str_maximized}")
        if str_maximized == "true":
            self.showMaximized()

        # splitter state
        val = qset.value("splitterSize")
        if val is not None:
            self._splitter.restoreState(val)

        qset.endGroup()
